README.md

```diff
@@ -130,83 +134,98 @@ const ipld = new Ipld({
+| `async Function` | `null` |
 
-Function to dynamically load an [IPLD Format](https://github.com/ipld/interface-ipld-format). It is passed a string `codec`, the multicodec of the IPLD format to load and a callback function to call when the format has been loaded. e.g.
+Function to dynamically load an [IPLD Format](https://github.com/ipld/interface-ipld-format). It is passed a `codec`, the multicodec code of the IPLD format to load and returns an IPLD Format implementation. For example:
 
 ```js
+const multicodec = require('multicodec')
 const ipld = new Ipld({
-  loadFormat (codec, callback) {
-    if (codec === 'git-raw') {
-      callback(null, require('ipld-git'))
+  async loadFormat (codec) {
+    if (codec === multicodec.GIT_RAW) {
+      return require('ipld-git')
     } else {
-      callback(new Error('unable to load format ' + codec))
+      throw new Error('unable to load format ' + multicodec.print[codec])
     }
   }
 })
 ```
 
-### `.put(node, options, callback)`
+### `.put(nodes, options)`
+
+> Stores the given IPLD Nodes of a recognized IPLD Format.
+
+ - `nodes` (`Iterable<Object>`): deserialized IPLD nodes that should be inserted.
+
+ - `options` is applied to any of the `nodes` and is an object with the following properties:
+   - `format` (`multicodec`, required): the multicodec of the format that IPLD Node should be encoded in.
+   - `hashAlg` (`multicodec`, default: hash algorithm of the given multicodec): the hashing algorithm that is used to calculate the CID.
+   - `version` (`number`, default: 1): the CID version to use.
+   - `onlyHash` (`boolean`, default: false): if true the serialized form of the IPLD Node will not be passed to the underlying block store.
+
+Returns an async iterator with the CIDs of the serialized IPLD Nodes.
+
+
+### `.resolve(cid, path)`
+
+> Retrieves IPLD Nodes along the `path` that is rooted at `cid`.
 
-> Store the given node of a recognized IPLD Format.
+ - `cid` (`CID`, required): the CID the resolving starts.
+ - `path` (`IPLD Path`, required): the path that should be resolved.
 
-`options` is an object that must contain one of the following combinations:
-- `cid` - the CID of the node
-- `[hashAlg]`, `[version]` and `format` - the hashAlg, version and the format that should be used to create the CID of the node. The
-`hashAlg` and `version` defaults to the default values for the `format`.
+Returns an async iterator of all the IPLD Nodes that were traversed during the path resolving. Every element is an object with these fields:
+  - `remainderPath` (`string`): the part of the path that wasn’t resolved yet.
+  - `value` (`*`): the value where the resolved path points to. If further traversing is possible, then the value is a CID object linking to another IPLD Node. If it was possible to fully resolve the path, `value` is the value the `path` points to. So if you need the CID of the IPLD Node you’re currently at, just take the `value` of the previously returned IPLD Node.
```

Comment:
Can we return the block instead of just the CID here? We have to parse the blocks to know if they resolve additional parts of the path anyway, so we have them, why not return them in the iteration instead of only returning the CID and forcing another lookup for the block data?

Reply:
Having an iterator as return value enables you to stop the iteration anywhere. If a Block is returned here, you would already be kind of in the next iteration. That Block might not be locally available. Hence I think returning just the CID makes sense here.